### PR TITLE
Add call snapshots

### PR DESCRIPTION
**Pull request by typfel**
We are adding call snapshots so that we don't have to query AVS about the previous
call state. This means we will no longer call the following methods:
```
wcall_get_state
wcall_is_video_call
wcall_iterate_state
```

The reason for this change is that the above mentioned methods do not return correct
values at all times. In addition to that AVS is not happy about how frequently we
call them atm.

### CODE REVIEW

File: Source/Calling/AVSWrapper.swift [hunk 1]
```diff
@@ -22,16 +22,13 @@ import avs
 
 public protocol AVSWrapperType {
     init(userId: UUID, clientId: String, observer: UnsafeMutableRawPointer?)
-    func callState(conversationId: UUID) -> CallState
```
Comment (basine) [inline Source/Calling/AVSWrapper.swift]: I wouldn't necessarily remove that since the methods are still there and might be useful some time in the future.

File: Source/Calling/WireCallCenterV3.swift [hunk 2]
```diff
@@ -449,40 +452,48 @@ public struct CallEvent {
...
+            callSnapshots[conversationId] = CallSnapshot(callState: callState, isVideo: video)
+            
+            participantSnapshots[conversationId] = VoiceChannelParticipantV3Snapshot(conversationId: conversationId,
+                                                                                     selfUserID: selfUserId,
+                                                                                     members: [CallMember(userId: userId!, audioEstablished: false)],
+                                                                                     initiator: userId)
+            break
+            
+        case .terminating(reason: let reason):
+            clearSnapshot(conversationId: conversationId)
```
Comment (basine) [inline Source/Calling/WireCallCenterV3.swift]: You don't want to clear the snapshot for every reason. e.g. when the user leaves the call but the call is still ongoing, we get an end reason .still_ongoing. We should only clear snapshots when we really know the call ended completely.

File: Source/Calling/AVSWrapper.swift (hunk 1 above)
Comment (basine) [inline Source/Calling/AVSWrapper.swift]: but for the sake of not reverting all your changes, it's ok if you remove it ;)

Comment (typfel) [inline Source/Calling/AVSWrapper.swift]: I wanted to be sure that it wasn't still  used somewhere

File: Source/Calling/WireCallCenterV3.swift (hunk 2 above)
Comment (typfel) [inline Source/Calling/WireCallCenterV3.swift]: This methods gets the `finalCallState` so the still-ongoing call state will never arrive here.

File: Tests/Source/Calling/WireCallCenterV3Tests.swift [hunk 3]
```diff
@@ -328,21 +333,31 @@ extension WireCallCenterV3Tests {
...
         
         // when
-        sut.mockAVSCallState = .incoming(video: false, shouldRing: true)
+        WireSyncEngine.incomingCallHandler(conversationId: conversationIdRef, messageTime: 0, userId: userIdRef, isVideoCall: 0, shouldRing: 1, contextRef: context)
+        XCTAssert(waitForAllGroupsToBeEmpty(withTimeout: 0.5))
         sut.rejectCall(conversationId: conversationId, isGroup: false)
         
         // then
-        XCTAssertNotEqual(sut.callState(conversationId: conversationId), .incoming(video: false, shouldRing: true))
+        XCTAssertNotEqual(sut.callState(conversationId: conversationId), .incoming(video: false, shouldRing: false))
```
Comment (typfel) [inline Tests/Source/Calling/WireCallCenterV3Tests.swift]: @basine this was bug before I think.

File: Source/Calling/WireCallCenterV3.swift (hunk 2 above)
Comment (basine) [inline Source/Calling/WireCallCenterV3.swift]: ah ok, you could just write `.terminating:` since you don't use reason, but it's ok like that.

File: Source/Calling/AVSWrapper.swift (hunk 1 above)
Comment (daehn) [inline Source/Calling/AVSWrapper.swift]: @typfel you could mark it as `@available (*, unavailable)`

Comment (typfel) [inline Source/Calling/AVSWrapper.swift]: @daehn true, next time ;)

File: Tests/Source/Calling/WireCallCenterV3Tests.swift (hunk 3 above)
Comment (basine) [inline Tests/Source/Calling/WireCallCenterV3Tests.swift]: I think in a 1on1 that's actually what we wanted to do, but AVS now handles it the same way as group calls: You get a still_ongoing followed by a cancelled afaik.

Comment (typfel) [inline Tests/Source/Calling/WireCallCenterV3Tests.swift]: Hmm ok, so we should show the call as still ongoing in the UI as well after ending a call?

Comment (basine) [inline Tests/Source/Calling/WireCallCenterV3Tests.swift]: tbh I have no idea, but I don't see a reason why we shouldn't. The call times out after a while anyway.

**Comment by basine** [comment]
@typfel As discussed, I would merge after the release.

File: Tests/Source/Calling/WireCallCenterV3Tests.swift (hunk 3 above)
Comment (typfel) [inline Tests/Source/Calling/WireCallCenterV3Tests.swift]: I guess bit it's a bit weird. You can't join the call again because it's in the process being shutdown...

Comment (basine) [inline Tests/Source/Calling/WireCallCenterV3Tests.swift]: No it's not - it only ends when it times out on the other side. The caller does not get notified the receiver rejected his call.

Comment (typfel) [inline Tests/Source/Calling/WireCallCenterV3Tests.swift]: Ah right, was thinking about close call. So I should revert this change then.

**Comment by daehn** [comment]
@typfel @basine can this be merged now?

**Comment by typfel** [comment]
@daehn @basine ah this thing, maybe it's better to wait until after the pilot release since we probably won't have time to fix any call issues.

**Comment by mikeger** [comment]
@typfel should we merge this one now?

**Comment by typfel** [comment]
@mikeger yes, know we should have some time to fix the fall out.

**Comment by typfel** [comment]
@mikeger wait still need do some fixes here.

**Comment by mikeger** [comment]
@basine let's merge this one?
